README.rst
This is an `ezjail-remote <https://github.com/tomster/ezjail-remote>`_ flavour that installs the `elektropost recipe <http://erdgeist.org/arts/software/elektropost/>`_, a fully functional IMAP and SMTP server complete with webmail, spam filtering and web based admin interface.

Note that both ezjail-remote and the recipe are `FreeBSD <http://freebsd.org/>`_ specific.



HISTORY.rst
0.2 - Planned
=============

 * Add sieve filtering and filter management


0.1 - Unreleased
================

 * Initial release covering the manual recipe at http://erdgeist.org/arts/software/elektropost/


ezjailremote/flavours/elektropost/__init__.py
import sys
from os import path

from fabric.api import sudo, task, put, puts, cd, env, warn
from fabric.contrib.files import upload_template
from OpenSSL import crypto

from ezjailremote.utils import is_ip


@task
def host_setup():
    """ Create additional loopback interface so that spamassasin can have its
    own IP to connect to (127.0.0.X is not available inside jails). """
    # create the interface and configure it:
    sudo('ifconfig lo1 create')
    sudo('ifconfig lo1 inet 192.168.0.2 netmask 0xfffffffe alias')
    sudo('ifconfig lo1 inet 192.168.0.3 netmask 0xffffffff alias')
    sudo('ifconfig lo1 inet 192.168.0.4 netmask 0xffffffff alias')
    sudo('ifconfig lo1 inet 192.168.0.5 netmask 0xffffffff alias')
    sudo('ifconfig lo1 inet 192.168.0.6 netmask 0xffffffff alias')
    # make it permanent:
    sudo('echo cloned_interfaces="lo1" >> /etc/rc.conf')
    sudo('echo ipv4_addrs_lo1="192.168.0.2-6/31" >> /etc/rc.conf')


@task
def setup(hostname, host_ip=None, pem_file=None):
    """ Sets up a fully functional mail server according to the elektropost
        instructions at http://erdgeist.org/arts/software/elektropost/

        Parameters:

        hostname: FQDN of the mail host, required

        host_ip: IP address of the host, required, if env.host is not an ip address

        pem_file: path to .pem file, will be used for IMAP and webmail, auto-generated,
            if none is  given
    """
    puts("running elektropost setup")
    # upload patches
    from ezjailremote.flavours import elektropost
    local_resource_dir = path.join(path.abspath(path.dirname(elektropost.__file__)))
    remote_patches_dir = "/tmp/ezjailremote.flavours.elektropost/"
    sudo("mkdir -p %s" % remote_patches_dir)
    put(path.join(local_resource_dir, 'patches'), remote_patches_dir, use_sudo=True)
    remote_patches_dir += 'patches'
    # upload ports options
    sudo("mkdir -p /var/db/ports/")
    put(path.join(local_resource_dir, 'server_root/var/db/ports/*'),
        "/var/db/ports/",
        use_sudo=True)

    if host_ip is None and is_ip.match(env['host']):
        host_ip = env['host']
    if host_ip is None:
        warn("No primary IP address specified! Either call using an ip address as "
            "host or explicitly pass one via host_ip")

    # Upload cert
    if pem_file is None:
        pem_file = path.join(local_resource_dir, 'servercert.pem')
        create_self_signed_cert(hostname, pem_file)
    elif not path.exists(pem_file):
        sys.exit("No such .pem '%s'" % pem_file)
    sudo("mkdir -p /var/qmail/control/")
    put(pem_file, "/var/qmail/control/servercert.pem", use_sudo=True)

    # Install qmail
    with cd("/usr/ports/mail/qmail-tls/"):
        sudo("make patch")

    with cd("/var/ports/basejail/usr/ports/mail/qmail-tls/work/qmail-1.03"):
        sudo("patch < %s" % path.join(remote_patches_dir, 'validrcptto.cdb.patch.new'))
        sudo("patch < %s" % path.join(remote_patches_dir, 'qmail-smtpd.c.privacy.patch'))

    with cd("/usr/ports/mail/qmail-tls/"):
        sudo("make install")
    sudo('echo "QMAIL_SLAVEPORT=tls" >> /etc/make.conf')

    # Configure qmail
    sudo("echo %s > /var/qmail/control/me" % hostname)
    sudo('ln -s /var/qmail/boot/qmail-smtpd.rcNG /usr/local/etc/rc.d/qmail-smtpd')
    sudo('ln -s /var/qmail/boot/maildir /usr/local/etc/rc.d/qmail')
    sudo('''echo 'qmailsmtpd_enable="YES"' >> /etc/rc.conf''')
    sudo('''echo 'qmailsmtpd_checkpassword="/usr/local/vpopmail/bin/vchkpw"' >> /etc/rc.conf''')
    sudo('cp %s/tcp.smtp /etc/' % remote_patches_dir)
    sudo('tcprules /etc/tcp.smtp.cdb /etc/tcp.smtp.tmp < /etc/tcp.smtp')

    # Install vpopmail
    with cd("/usr/ports/mail/vpopmail"):
        sudo("make install")
    sudo("chown vpopmail:vchkpw /usr/local/vpopmail")
    sudo("chmod u+s ~vpopmail/bin/vchkpw")
    sudo("pw user mod vpopmail -s /bin/sh")
    # Configure vpopmail
    sudo("echo %s > /usr/local/vpopmail/etc/defaultdomain" % hostname)

    # Install dovecot
    with cd("/usr/ports/mail/dovecot"):
        sudo("make install")
    sudo('''echo 'dovecot_enable="YES"' >> /etc/rc.conf''')

    # Configure dovecot
    upload_template(path.join(local_resource_dir, 'dovecot.conf'),
        '/usr/local/etc/dovecot.conf',
        context=dict(
            hostname=hostname,
            dovecot_ip=host_ip),
        backup=False,
        use_sudo=True)

    # Install lighty
    with cd("/usr/ports/www/lightttpd"):
        sudo("make install")
    sudo('''echo 'lighttpd_enable="YES"' >> /etc/rc.conf''')
    sudo('touch /var/log/lighttpd.error.log')
    sudo('chown www:www /var/log/lighttpd.error.log')
    sudo('touch /var/log/lighttpd.access.log')
    sudo('chown www:www /var/log/lighttpd.access.log')
    sudo('mkdir /var/run/lighttpd/')
    sudo('chown www:www /var/run/lighttpd')

    # Configure lighty
    sudo("mkdir /usr/local/etc/lighttpd/")
    upload_template(path.join(local_resource_dir, 'lighttpd.conf'),
        '/usr/local/etc/lighttpd/lighttpd.conf',
        context=dict(
            hostname=hostname,
            httpd_ip=host_ip),
        backup=False,
        use_sudo=True)
    sudo("mkdir -p /usr/local/www/data")
    put(path.join(local_resource_dir, 'data/*.*'),
        "/usr/local/www/data/", use_sudo=True)
    upload_template(path.join(local_resource_dir, 'data/index.html'),
        '/usr/local/www/data/index.html',
        context=dict(
            hostname=hostname,
            httpd_ip=host_ip),
        backup=False,
        use_sudo=True)

    # Install squirrelmail
    with cd("/usr/ports/mail/squirrelmail"):
        sudo("make install")

    # Configure squirrelmail
    upload_template(path.join(local_resource_dir, 'config.php'),
        '/usr/local/www/squirrelmail/config/config.php',
        context=dict(
            hostname=hostname,
            httpd_ip=host_ip),
        backup=False,
        use_sudo=True)

    # Install qmailadmin / ezmlm-idx
    with cd("/usr/ports/mail/qmailadmin"):
        sudo('make install WITH_SPAM_DETECTION=TRUE SPAM_COMMAND="| /usr/local/bin/spamc -f | /usr/local/bin/maildrop" CGIBINDIR=www/squirrelmail/cgi-bin CGIBINSUBDIR= WEBDATADIR=www/squirrelmail WEBDATASUBDIR=qmailadmin')

    # Install qmailadmin plugin for squirrelmail
    with cd("/usr/ports/mail/squirrelmail-qmailadmin_login-plugin"):
        sudo("make install")

    # Install maildrop
    with cd("/usr/ports/mail/maildrop"):
        sudo("make install")

    # Install the maildrop spam sort magic
    sudo("mv %s /usr/local/etc/maildroprc" % path.join(remote_patches_dir, 'maildroprc'))

    # Install spamassassin
    with cd("/usr/ports/mail/p5-Mail-SpamAssassin"):
        sudo("make install")
    sudo('mkdir -p /usr/local/etc/mail/spamassassin')
    sudo("echo '-d 192.168.0.2' > /usr/local/etc/mail/spamassassin/spamc.conf")
    # TODO: answer YES when prompted whether to download new rules


def create_self_signed_cert(hostname, pem_file):
    """ based on http://skippylovesmalorie.wordpress.com/2010/02/12/how-to-generate-a-self-signed-certificate-using-pyopenssl/
    """
    # create a key pair
    pkey = crypto.PKey()
    pkey.generate_key(crypto.TYPE_RSA, 1024)

    # create a minimal self-signed cert
    cert = crypto.X509()
    cert.get_subject().CN = hostname
    cert.set_serial_number(1000)
    cert.gmtime_adj_notBefore(0)
    cert.gmtime_adj_notAfter(365 * 24 * 60 * 60)
    cert.set_issuer(cert.get_subject())
    cert.set_pubkey(pkey)
    cert.sign(pkey, 'sha1')

    open(pem_file, "wt").write(
        crypto.dump_privatekey(crypto.FILETYPE_PEM, pkey) +
        crypto.dump_certificate(crypto.FILETYPE_PEM, cert))


ezjailremote/flavours/elektropost/config.php
<?php

/**
 * SquirrelMail Configuration File
 */

global $version;
$config_version = '1.4.0';
$config_use_color = 2;

$org_name      = "%(hostname)s";
$org_logo      = SM_PATH . 'images/sm_logo.png';
$org_logo_width  = '388';
$org_logo_height = '90';
$org_title     = "elektropost $version";
$signout_page  = '';
$frame_top     = '_top';

$provider_uri     = 'http://erdgeist.org/arts/software/elektropost/';
$provider_name     = 'Netzkooperative Berlin';
$motd = "";

$squirrelmail_default_language = 'en_US';
$default_charset       = 'iso-8859-1';
$lossy_encoding        = false;

$domain                 = '%(hostname)s';
$imapServerAddress      = '%(hostname)s';
$imapPort               = 143;
$useSendmail            = false;
$smtpServerAddress      = '%(hostname)s';
$smtpPort               = 25;
$sendmail_path          = '/usr/sbin/sendmail';
$sendmail_args          = '-i -t';
$pop_before_smtp        = false;
$pop_before_smtp_host   = '';
$imap_server_type       = 'other';
$invert_time            = false;
$optional_delimiter     = 'detect';
$encode_header_key      = 'plasmakocher';

$default_folder_prefix          = '';
$trash_folder                   = 'INBOX.Trash';
$sent_folder                    = 'INBOX.Sent';
$draft_folder                   = 'INBOX.Drafts';
$default_move_to_trash          = true;
$default_move_to_sent           = true;
$default_save_as_draft          = true;
$show_prefix_option             = false;
$list_special_folders_first     = true;
$use_special_folder_color       = true;
$auto_expunge                   = true;
$default_sub_of_inbox           = true;
$show_contain_subfolders_option = false;
$default_unseen_notify          = 2;
$default_unseen_type            = 1;
$auto_create_special            = true;
$delete_folder                  = false;
$noselect_fix_enable            = false;

$data_dir                 = '/var/spool/squirrelmail/pref/';
$attachment_dir           = '/var/spool/squirrelmail/attach/';
$dir_hash_level           = 0;
$default_left_size        = '150';
$force_username_lowercase = false;
$default_use_priority     = true;
$hide_sm_attributions     = false;
$default_use_mdn          = true;
$edit_identity            = true;
$edit_name                = true;
$hide_auth_header         = false;
$allow_thread_sort        = false;
$allow_server_sort        = false;
$allow_charset_search     = true;
$uid_support              = true;

$plugins[0] = 'qmailadmin_login';

$theme_css = '';
$theme_default = 0;
$theme[0]['PATH'] = SM_PATH . 'themes/default_theme.php';
$theme[0]['NAME'] = 'Default';
$theme[1]['PATH'] = SM_PATH . 'themes/plain_blue_theme.php';
$theme[1]['NAME'] = 'Plain Blue';
$theme[2]['PATH'] = SM_PATH . 'themes/sandstorm_theme.php';
$theme[2]['NAME'] = 'Sand Storm';
$theme[3]['PATH'] = SM_PATH . 'themes/deepocean_theme.php';
$theme[3]['NAME'] = 'Deep Ocean';
$theme[4]['PATH'] = SM_PATH . 'themes/slashdot_theme.php';
$theme[4]['NAME'] = 'Slashdot';
$theme[5]['PATH'] = SM_PATH . 'themes/purple_theme.php';
$theme[5]['NAME'] = 'Purple';
$theme[6]['PATH'] = SM_PATH . 'themes/forest_theme.php';
$theme[6]['NAME'] = 'Forest';
$theme[7]['PATH'] = SM_PATH . 'themes/ice_theme.php';
$theme[7]['NAME'] = 'Ice';
$theme[8]['PATH'] = SM_PATH . 'themes/seaspray_theme.php';
$theme[8]['NAME'] = 'Sea Spray';
$theme[9]['PATH'] = SM_PATH . 'themes/bluesteel_theme.php';
$theme[9]['NAME'] = 'Blue Steel';
$theme[10]['PATH'] = SM_PATH . 'themes/dark_grey_theme.php';
$theme[10]['NAME'] = 'Dark Grey';
$theme[11]['PATH'] = SM_PATH . 'themes/high_contrast_theme.php';
$theme[11]['NAME'] = 'High Contrast';
$theme[12]['PATH'] = SM_PATH . 'themes/black_bean_burrito_theme.php';
$theme[12]['NAME'] = 'Black Bean Burrito';
$theme[13]['PATH'] = SM_PATH . 'themes/servery_theme.php';
$theme[13]['NAME'] = 'Servery';
$theme[14]['PATH'] = SM_PATH . 'themes/maize_theme.php';
$theme[14]['NAME'] = 'Maize';
$theme[15]['PATH'] = SM_PATH . 'themes/bluesnews_theme.php';
$theme[15]['NAME'] = 'BluesNews';
$theme[16]['PATH'] = SM_PATH . 'themes/deepocean2_theme.php';
$theme[16]['NAME'] = 'Deep Ocean 2';
$theme[17]['PATH'] = SM_PATH . 'themes/blue_grey_theme.php';
$theme[17]['NAME'] = 'Blue Grey';
$theme[18]['PATH'] = SM_PATH . 'themes/dompie_theme.php';
$theme[18]['NAME'] = 'Dompie';
$theme[19]['PATH'] = SM_PATH . 'themes/methodical_theme.php';
$theme[19]['NAME'] = 'Methodical';
$theme[20]['PATH'] = SM_PATH . 'themes/greenhouse_effect.php';
$theme[20]['NAME'] = 'Greenhouse Effect (Changes)';
$theme[21]['PATH'] = SM_PATH . 'themes/in_the_pink.php';
$theme[21]['NAME'] = 'In The Pink (Changes)';
$theme[22]['PATH'] = SM_PATH . 'themes/kind_of_blue.php';
$theme[22]['NAME'] = 'Kind of Blue (Changes)';
$theme[23]['PATH'] = SM_PATH . 'themes/monostochastic.php';
$theme[23]['NAME'] = 'Monostochastic (Changes)';
$theme[24]['PATH'] = SM_PATH . 'themes/shades_of_grey.php';
$theme[24]['NAME'] = 'Shades of Grey (Changes)';
$theme[25]['PATH'] = SM_PATH . 'themes/spice_of_life.php';
$theme[25]['NAME'] = 'Spice of Life (Changes)';
$theme[26]['PATH'] = SM_PATH . 'themes/spice_of_life_lite.php';
$theme[26]['NAME'] = 'Spice of Life - Lite (Changes)';
$theme[27]['PATH'] = SM_PATH . 'themes/spice_of_life_dark.php';
$theme[27]['NAME'] = 'Spice of Life - Dark (Changes)';
$theme[28]['PATH'] = SM_PATH . 'themes/christmas.php';
$theme[28]['NAME'] = 'Holiday - Christmas';
$theme[29]['PATH'] = SM_PATH . 'themes/darkness.php';
$theme[29]['NAME'] = 'Darkness (Changes)';
$theme[30]['PATH'] = SM_PATH . 'themes/random.php';
$theme[30]['NAME'] = 'Random (Changes every login)';
$theme[31]['PATH'] = SM_PATH . 'themes/midnight.php';
$theme[31]['NAME'] = 'Midnight';
$theme[32]['PATH'] = SM_PATH . 'themes/alien_glow.php';
$theme[32]['NAME'] = 'Alien Glow';
$theme[33]['PATH'] = SM_PATH . 'themes/dark_green.php';
$theme[33]['NAME'] = 'Dark Green';
$theme[34]['PATH'] = SM_PATH . 'themes/penguin.php';
$theme[34]['NAME'] = 'Penguin';
$theme[35]['PATH'] = SM_PATH . 'themes/minimal_bw.php';
$theme[35]['NAME'] = 'Minimal BW';
$theme[36]['PATH'] = SM_PATH . 'themes/redmond.php';
$theme[36]['NAME'] = 'Redmond';
$theme[37]['PATH'] = SM_PATH . 'themes/netstyle_theme.php';
$theme[37]['NAME'] = 'Net Style';
$theme[38]['PATH'] = SM_PATH . 'themes/silver_steel_theme.php';
$theme[38]['NAME'] = 'Silver Steel';
$theme[39]['PATH'] = SM_PATH . 'themes/simple_green_theme.php';
$theme[39]['NAME'] = 'Simple Green';
$theme[40]['PATH'] = SM_PATH . 'themes/wood_theme.php';
$theme[40]['NAME'] = 'Wood';
$theme[41]['PATH'] = SM_PATH . 'themes/bluesome.php';
$theme[41]['NAME'] = 'Bluesome';
$theme[42]['PATH'] = SM_PATH . 'themes/simple_green2.php';
$theme[42]['NAME'] = 'Simple Green 2';
$theme[43]['PATH'] = SM_PATH . 'themes/simple_purple.php';
$theme[43]['NAME'] = 'Simple Purple';
$theme[44]['PATH'] = SM_PATH . 'themes/autumn.php';
$theme[44]['NAME'] = 'Autumn';
$theme[45]['PATH'] = SM_PATH . 'themes/autumn2.php';
$theme[45]['NAME'] = 'Autumn 2';
$theme[46]['PATH'] = SM_PATH . 'themes/blue_on_blue.php';
$theme[46]['NAME'] = 'Blue on Blue';
$theme[47]['PATH'] = SM_PATH . 'themes/classic_blue.php';
$theme[47]['NAME'] = 'Classic Blue';
$theme[48]['PATH'] = SM_PATH . 'themes/classic_blue2.php';
$theme[48]['NAME'] = 'Classic Blue 2';
$theme[49]['PATH'] = SM_PATH . 'themes/powder_blue.php';
$theme[49]['NAME'] = 'Powder Blue';
$theme[50]['PATH'] = SM_PATH . 'themes/techno_blue.php';
$theme[50]['NAME'] = 'Techno Blue';
$theme[51]['PATH'] = SM_PATH . 'themes/turquoise.php';
$theme[51]['NAME'] = 'Turquoise';

$default_use_javascript_addr_book = false;
$abook_global_file = '';
$abook_global_file_writeable = false;
$abook_global_file_listing = true;
$abook_file_line_length = 2048;

$addrbook_dsn = '';
$addrbook_table = 'address';

$prefs_dsn = '';
$prefs_table = 'userprefs';
$prefs_user_field = 'user';
$prefs_key_field = 'prefkey';
$prefs_val_field = 'prefval';
$addrbook_global_dsn = '';
$addrbook_global_table = 'global_abook';
$addrbook_global_writeable = false;
$addrbook_global_listing = false;

$no_list_for_subscribe = false;
$smtp_auth_mech = 'login';
$imap_auth_mech = 'login';
$smtp_sitewide_user = '';
$smtp_sitewide_pass = '';
$use_imap_tls = false;
$use_smtp_tls = false;
$session_name = 'SQMSESSID';
$only_secure_cookies   = true;

$config_location_base     = '';

@include SM_PATH . 'config/config_local.php';

